### DISCUSSION TITLE
Discrepancies between grid operator data and amsreader data in norwegian tariff level data

### DISCUSSION THREAD

**Discussion by oivindoh** [UtilitechAS/amsreader-firmware#436] [General]
I try to live on the edge when it comes to the 10kW limit (because 5kW is impossible with two EVs that do a fair bit of driving, and 15kW seems a waste), so for a couple of months I've paid more attention to this data.

This month, I've accidentally sent myself over the limit by a hair, annoyingly enough, and I see AMS reader is reporting 10.20, 10.26 and 9.57 as peaks on the 9th, 3rd and 10th respectively. This puts me annoyingly close to not being over.

However, the grid operator has other numbers for these peaks;

- On the 10th, the grid operator is in agreement with a 9,578 hour the second hour of that day.
- On the 9th, they have a peak of 10,203 registered on me from hour 4. This is in line with ams reader.
- On the 3rd, the grid operator has a peak of 10,403 on me for the first hour, with the next highest hour being the second at 9,035. There is no sign of the 10,26 hour, though I assume it's the same hour the grid operator has listed at 10,403.

The grid operator also lists a peak on the 2nd at 10,652 (the last hour) which the ams reader appears to have completely missed (bringing my month average up to 10,419). I guess there must have been some sort of connectivity issue around midnight - I'm not familiar with the data the reader receives, but does it not also receive a copy of the full reports sent to the grid operator, so that it could maybe retroactively update hourly consumption numbers if these differ?

**Comment by ArnieO** [comment]
Sometimes the remote reading of the meter fails, for shorter or longer periods. In such cases the grid company data point is stipulated (not measured). If you log in to "Min side" on https://elhub.no/ you can download the registered data. You will see an indication on each measurement whether any data point is stipulated. 
(We tried to ask Elhub on Twitter how they do the stipulation, but no answer: https://twitter.com/amsleser/status/1617965710009069569?s=20 )

Can you check the data point where you see a discrepancy?

**Reply by oivindoh** [reply]
They appear to be measured according to the csv I got from elhub (all of the february data there is "Avlest":

| start | end | kWh | stip |
| --- | --- | --- | --- |
| 02.02.2023 21:00 | 02.02.2023 22:00 | 3,196 | Avlest |
| 02.02.2023 22:00 | 02.02.2023 23:00 | 8,812 | Avlest |
| 02.02.2023 23:00 | 03.02.2023 00:00 | 10,652 | Avlest |
| 03.02.2023 00:00 | 03.02.2023 01:00 | 10,403 | Avlest |
| 03.02.2023 01:00 | 03.02.2023 02:00 | 9,035 | Avlest |
| 03.02.2023 02:00 | 03.02.2023 03:00 | 6,213 | Avlest |

**Reply by oivindoh** [reply]
A bit dismayed to see elhub use twitter the wrong way (just like a megaphone)

**Reply by ArnieO** [reply]
> A bit dismayed to see elhub use twitter the wrong way (just like a megaphone)

Yep - disappointing, but sadly not unique.

**Comment by ThorAlex87** [comment]
I also have minor discrepancies, but they are all within 0,02kwh so they are pretty negligible and well within the margins in my setup. Yours sound a bit more problematic. My power company (Gudbrandsdals Energi) shows stipulated usage in the app, and so far that has always showed lower peaks than what was measured by the pow-u.

**Comment by ArnieO** [comment]
> I'm not familiar with the data the reader receives, but does it not also receive a copy of the full reports sent to the grid operator, so that it could maybe retroactively update hourly consumption numbers if these differ?

OK, so your deviations are difficult to explain, could be related to whole-hour datagrams were missed for some reason (reboot of device? I can only speculate).

But I can explain the data flow a bit. 

Firstly - I have not seen published information regarding the communication on the grid operator's network. I assume there is a buffer in the power meter, in case the wireless communication is lost for a shorter or longer period. I have no idea how large such a buffer could be. 

The data that is pushed on the HAN port is:

- Every X second (depends on meter brand): Power [W] and some other data, but not energy [Wh]. 
- Shortly after each whole hour, a larger datagram arrives, where also energy is included, showing total consumption since meter was installed. 

The firmware calculates hourly consumption by subtracting current and previous accumulated energy [Wh].

The real-time counter shown on the main page of our firmware is derived by integrating/accumulating time * power from the meter, to get a running estimate. That estimate is then validated/updated when the whole-hour datagram arrives. 

If one such whole-hour datagram is for some reason not captured by the reader (maybe it rebooted for some reason), the value may be slightly wrong. We do our best to ensure the firmware is stable, so such reboots should not be frequent - but they may still happen. We have also seen that some ESP devices reeboot from time to time (but infrequently) - with identical firmware as an other identical device. We believe that to be some kind of hardware error in them, but so marginal that it is not possible to detect it during our testing (done before shipping Pow-U/K/P1 devices).
